src/main/java/com/theelixrlabs/healthcare/service/DoctorPatientAssignmentService.java

```diff
@@ -0,0 +1,97 @@
+    private final PatientRepository patientRepository;
+    private final DoctorRepository doctorRepository;
+    private final MessageSource messageSource;
+
+    public DoctorPatientAssignmentService(DoctorPatientAssignmentRepository doctorPatientAssignmentRepository,
+                                          PatientRepository patientRepository, DoctorRepository doctorRepository,
+                                          MessageSource messageSource) {
+        this.doctorPatientAssignmentRepository = doctorPatientAssignmentRepository;
+        this.patientRepository = patientRepository;
+        this.doctorRepository = doctorRepository;
+        this.messageSource = messageSource;
+    }
+
+
+    private void validateDoctorPatientAssignmentDTO(DoctorPatientAssignmentDTO doctorPatientAssignmentDTO) throws CustomException {
+        UUID doctorId, patientId;
+        try {
+            doctorId = UUID.fromString(doctorPatientAssignmentDTO.getDoctorID());
+        } catch (IllegalArgumentException illegalArgumentException) {
+            throw new CustomException(DoctorPatientAssignmentConstants.DOCTOR_ID_INVALID_KEY, messageSource);
+        }
+        try {
+            patientId = UUID.fromString(doctorPatientAssignmentDTO.getPatientID());
+        } catch (IllegalArgumentException illegalArgumentException) {
+            throw new CustomException(DoctorPatientAssignmentConstants.PATIENT_ID_INVALID_KEY, messageSource);
+        }
+        if (!doctorRepository.existsById(doctorId))
+            throw new CustomException(DoctorPatientAssignmentConstants.DOCTOR_NOT_FOUND_KEY, messageSource);
+        if (!patientRepository.existsById(patientId))
+            throw new CustomException(DoctorPatientAssignmentConstants.PATIENT_NOT_FOUND_KEY, messageSource);
+    }
+
+    /**
+     * Checks if doctor is previously assigned to same patient and if not then saves the doctorPatientAssignmentDTO
+     *
+     * @param doctorPatientAssignmentDTO DTO object containing doctorId and patientId
+     * @return DoctorPatientAssignmentDTO
+     * @throws CustomException if any exception occurs
+     */
+    public DoctorPatientAssignmentDTO assignDoctorToPatient(DoctorPatientAssignmentDTO doctorPatientAssignmentDTO) throws CustomException {
+        validateDoctorPatientAssignmentDTO(doctorPatientAssignmentDTO);
+        List<DoctorPatientAssignmentModel> doctorPatientAssignmentList =
+                doctorPatientAssignmentRepository.findByDoctorIDAndPatientID
+                        (UUID.fromString(doctorPatientAssignmentDTO.getDoctorID()),
+                                UUID.fromString(doctorPatientAssignmentDTO.getPatientID()));
+        if (!doctorPatientAssignmentList.isEmpty()) {
+            for (DoctorPatientAssignmentModel doctorPatientAssignmentModel : doctorPatientAssignmentList) {
+                if (doctorPatientAssignmentModel.getDateOfUnassignment() == null) {
```

Comment:
Instead of checking the un-assignment here, Directly you can filter it from Repo Method. This will avoid the usage of for loop and improve the performance.

Reply:
Changed the repository method so that it will return the required object